src/Chapter4.hs

```diff
@@ -673,6 +712,48 @@ Specifically,
  ❃ Implement the function to convert Tree to list
 -}
 
+data Tree a = None | Node { nodeValue :: a
+                          , nodeLeft :: Tree a
+                          , nodeRight :: Tree a } deriving (Show)
+
+instance Functor Tree where
+  fmap :: (a -> b) -> Tree a -> Tree b
+  fmap _ None = None
+  fmap f (Node {..}) = Node { nodeValue = f nodeValue
+                            , nodeLeft = fmap f nodeLeft
+                            , nodeRight = fmap f nodeRight }
+
+tree :: Tree Integer
+tree = Node { nodeValue = 10
+            , nodeLeft = Node { nodeValue = 5
+                              , nodeLeft = Node { nodeValue = 1
+                                                , nodeLeft = None
+                                                , nodeRight = None }
+                              , nodeRight = None }
+            , nodeRight = Node { nodeValue = 15
+                               , nodeLeft = None
+                               , nodeRight = None } }
+
+reverseTree :: Tree a -> Tree a
+reverseTree None = None
+reverseTree node@Node {..} = node { nodeLeft = reverseTree nodeRight
+                                  , nodeRight = reverseTree nodeLeft }
+
+-- reverseTree tree
+-- Node { nodeValue = 10
+--      , nodeLeft = Node { nodeValue = 15
+--                        , nodeLeft = None
+--                        , nodeRight = None }
+--      , nodeRight = Node { nodeValue = 5
+--                         , nodeLeft = None
+--                         , nodeRight = Node { nodeValue = 1
+--                                            , nodeLeft = None
+--                                           , nodeRight = None } } }
+
+treeToList :: Tree a -> [a]
+treeToList None = []
+treeToList Node {..} = (treeToList nodeLeft) ++ [nodeValue] ++ (treeToList nodeRight)
+-- You said that ++ is slow, i know, but i can't understand how to avoid it in this case.
```

Comment:
Yeah, it is not possible to avoid it in here 🙂 This is the correct implementation 👍🏼